docs/src/explanations/dpi.md

```diff
@@ -0,0 +1,143 @@
+
+### Argument Types
+
+* Operand and result types must be passive.
+* A vector is lowered to an *unpacked* *open* array type, e.g., `a: Vec<4, UInt>` to `byte a []`.
+* A bundle is lowered to a packed struct.
+* Integer types are lowered into 2-state types.
+Small integer types (< 64 bit) must be compatible with C-types and arguments are passed by value. Users are required to use specific integer types for small integers shown in the table below. Large integers are lowered to bit and passed by reference.
+
+
+| Width | Verilog Type | Argument Passing Modes |
+| ----- | ------------ | ---------------------- |
+| 1     | bit          | value                  |
+| 8     | byte         | value                  |
+| 16    | shortint     | value                  |
+| 32    | int          | value                  |
+| 64    | longint      | value                  |
+| > 64  | bit [w-1:0]  | reference              |
+
+### Function Types
+The type of DPI object you need depends on the characteristics of your C++ function:
+
+* Return Type
+  * For functions that don't return a value (like hello), use `DPIVoidFunctionImport`.
+  * For functions with a return type (e.g., integer addition), use `DPINonVoidFunctionImport[T]`, where `T` is the return type.
+  * The output argument must be the last argument in DPI fuction.
+* Clocked vs. Unclocked:
+  * Clocked: The function call is evaluated at the associated clock's positive edge. By default, Chisel uses the current module's clock. For custom clocks, use `withClocked(clock) {..}`. If the function has a return value, it will be available in the next clock cycle.
+  * Unclocked: The function call is evaluated immediately.
+
+## Example: Adding Two Numbers
+Here's an example of a DPI function that calculates the sum of two numbers:
+
+```c++
+extern "C" void add(int lhs, int rhs, int* result)
+{
+    *result = lhs + rhs;
+}
+```
+
+```scala
+object Add extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "add"
+  override val ret = UInt(32.W)
+  override val clocked = false
+  override val inputNames = Some(Seq("lhs", "rhs"))
+  override val outputName = Some("result")
+  final def apply(lhs: UInt, rhs: UInt): UInt = super.call(lhs, rhs)
+}
+
+val result = Add(2.U(32.W), 3.U(32.W))
+```
+
+Explanation:
+
+* `Add` inherits from `DPINonVoidFunctionImport[UInt]` because it returns a 32-bit unsigned integer.
+* `ret` specifies the return type.
+* `clocked` indicates that this is a clocked function call.
+* `inputNames` and `outputName` provide optional names for the function's arguments and return value (these are just for Verilog readability).
+
+## Example: Sum of an array
+Chisel vectors are converted into SystemVerilog open arrays when used with DPI. Since memory layout can vary between simulators, it's recommended to use `svSize` and `svGetBitArrElemVecVal` to access array elements.
+
+```c++
+extern "C" void sum(const svOpenArrayHandle array, int* result) {
```

Comment:
What is `svOpenArrayHandle`?

I guess, more generally, what libraries does Chisel depend on in the C++ DPI? We can't simply assume that a general Chisel user will have whatever library exposes `DPI.h` or whatever it is. 

It would be good to have a section at the very start that gives the C++ prerequisites.